zipline/pipeline/loaders/blaze.py

```diff
@@ -0,0 +1,904 @@
+        # This is a single column. Record which column we are to return
+        # but create the entire dataset.
+        single_column = rename = expr._name
+        field_hit = False
+        if not isinstance(expr, traversable_nodes):
+            raise TypeError(
+                "expression '%s' was array-like but not a simple field of"
+                " some larger table" % str(expr),
+            )
+        while isinstance(expr, traversable_nodes):
+            if isinstance(expr, bz.expr.Field):
+                if not field_hit:
+                    field_hit = True
+                else:
+                    break
+            rename = expr._name
+            expr = expr._child
+        dataset_expr = expr.relabel({rename: single_column})
+    else:
+        dataset_expr = expr
+
+    measure = dataset_expr.dshape.measure
+    if not isrecord(measure) or AD_FIELD_NAME not in measure.names:
+        raise TypeError(
+            "The dataset must be a collection of records with at least an"
+            " '{ad}' field. Fields provided: '{fields}'\nhint: maybe you need"
+            " to use `relabel` to change your field names".format(
+                ad=AD_FIELD_NAME,
+                fields=measure,
+            ),
+        )
+    _check_datetime_field(AD_FIELD_NAME, measure)
+    dataset_expr, deltas = _ensure_timestamp_field(dataset_expr, deltas)
+
+    if deltas is not None and (sorted(deltas.dshape.measure.fields) !=
+                               sorted(measure.fields)):
+        raise TypeError(
+            'baseline measure != deltas measure:\n%s != %s' % (
+                measure,
+                deltas.dshape.measure,
+            ),
+        )
+
+    # Ensure that we have a data resource to execute the query against.
+    _check_resources('dataset_expr', dataset_expr, resources)
+    _check_resources('deltas', deltas, resources)
+
+    # Create or retrieve the Pipeline API dataset.
+    ds = new_dataset(dataset_expr, deltas)
+    # Register our new dataset with the loader.
+    (loader if loader is not None else global_loader)[ds] = ExprData(
+        dataset_expr,
+        deltas,
+        resources,
+    )
+    if single_column is not None:
+        # We were passed a single column, extract and return it.
+        return getattr(ds, single_column)
+    return ds
+
+
+getdataset = op.attrgetter('dataset')
+dataset_name = op.attrgetter('name')
+
+
+def inline_novel_deltas(baseline, deltas, dates):
+    """Inline any deltas into the baseline set that would have changed our most
+    recently known value.
+
+    Parameters
+    ----------
+    baseline : pd.DataFrame
+        The first known values.
+    deltas : pd.DataFrame
+        Overwrites to the baseline data.
+    dates : pd.DatetimeIndex
+        The dates requested by the loader.
+
+    Returns
+    -------
+    new_baseline : pd.DataFrame
+        The new baseline data with novel deltas inserted.
+    """
+    get_indexes = dates.searchsorted
```

Comment:
I think this can create duplicate entries if there's a "novel delta" on a date for which we have a baseline entry.